### PR TITLE
feat: implement MCP server with Server stdio/sse transport

### PR DESCRIPTION
**Pull request by tzolov**
MCP Server:

- Add McpServer factory with builder pattern for flexible configuration
- Add McpAsyncServer for non-blocking operations with reactive support
- Add McpSyncServer as synchronous wrapper around async implementation
- Implement tool management with runtime addition/removal capabilities
- Add support for server capabilities and implementation info

Transport Layer:

- Add SseServerTransport with SSE endpoints and message routing
- Add StdioServerTransport for process-based communication
- Move client transports to dedicated transport package
- Improve transport error handling and graceful shutdown

Documentation:

- Add detailed MCP reference documentation
- Client/Server architecture and interactions
- Transport implementations and protocols
- Session management and message handling
- Error handling and recovery strategies
- Add architecture diagrams showing component relationships
- Add API documentation with usage examples

Testing:

- Add extensive test coverage for server components
- Add transport-specific test suites
- Add integration tests for client-server interactions

BREAKING CHANGES:

Transport Package Reorganization:

- StdioClientTransport moved from o.s.ai.mcp.client.stdio to o.s.ai.mcp.client.transport
- SseClientTransport moved from o.s.ai.mcp.client.sse to o.s.ai.mcp.client.transport
- ServerParameters moved from o.s.ai.mcp.client.stdio to o.s.ai.mcp.client.transport Migration: Update import statements to use new package paths

Transport API Changes:

- StdioClientTransport.errorHandler changed to private with setter method Migration: Use setErrorHandler() method instead of direct field access

McpSchema Changes:

- Added new ServerCapabilities builder pattern Migration: Use new builder pattern instead of direct constructor:
 - Before: new ServerCapabilities(experimental, logging, prompts, resources, tools)
 - After: ServerCapabilities.builder() .experimental(experimental) .logging(logging) .prompts(listChanged) .resources(subscribe, listChanged) .tools(listChanged) .build()

### CODE REVIEW

File: mcp/src/main/java/org/springframework/ai/mcp/server/McpServer.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+ * <pre>{@code
+ * McpServer.using(transport)
+ * 		.serverInfo(new McpSchema.Implementation("my-server", "1.0.0"))
+ * 		.tool(new MyToolHandler())
+ * 		.async(); // or .sync()
+ * }</pre>
+ *
+ * @author Christian Tzolov
+ */
+public class McpServer {
```
Comment (kpavlov) [inline mcp/src/main/java/org/springframework/ai/mcp/server/McpServer.java:57]: I see this class might be converted to `interface` without any other api changes. Same for `org.springframework.ai.mcp.client.McpClient` class.
Anyway, the Builder fluent interface looks very nice and clean

File: docs/ref-client.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
+# MCP Client Reference
+
+## McpClient Factory
+
+`org.springframework.ai.mcp.client.McpClient`
+
+The `McpClient` class serves as a factory for creating Model Context Protocol (MCP) clients. It provides the main entry point for establishing connections with MCP servers, offering both synchronous and asynchronous client implementations.
```
Comment (kpavlov) [inline docs/ref-client.md:7]: nit: naming is a little bit confusing: I would expect a factory to have a `Factory` suffix. Same for McpServer

File: spring-ai-mcp-sample/src/main/resources/application.properties [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Server port
+# server.port=8080
+
+# Logging
+# logging.level.org.springframework.ai.mcp=DEBUG
```
Comment (kpavlov) [inline spring-ai-mcp-sample/src/main/resources/application.properties:5]:
```suggestion
logging.level.org.springframework.ai.mcp=INFO
```
It make sense to show something in the logs

File: mcp/src/test/java/org/springframework/ai/mcp/attic/DemoServer.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+	public static void main(String[] args) {
+		SseServerTransport transport = new SseServerTransport(new ObjectMapper(), "/mcp/message");
+
+		var mcpServer = McpServer.using(transport)
+			.serverInfo("Weather Forecast", "1.0.0")
+			.tool(new McpSchema.Tool("weather", "Weather forecast tool by location", Map.of("city", "String")),
+					(arguments) -> {
+						String city = (String) arguments.get("city");
+						return new CallToolResult(List.of(), false);
```
Comment (kpavlov) [inline mcp/src/test/java/org/springframework/ai/mcp/attic/DemoServer.java:50]: Implementation is missing. Hardcoded result, like in McpServerConfig would be good enough.
```java
TextContent content = new TextContent("Weather forecast for " + city + " is sunny");
return new CallToolResult(List.of(content), false);
```
Or should this class be removed completely in favour of McpServerApplication?

File: docs/ref-transport.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+**Parameters:**
+- `message`: The JSON-RPC message to be sent to the server
+
+**Returns:**
+- A `Mono<Void>` that completes when the message has been sent
+
+#### Close Gracefully
+```java
+Mono<Void> closeGracefully()
```
Comment (kpavlov) [inline docs/ref-transport.md:50]: Does it make sense to provide also another method with timeout?

File: docs/ref-transport.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+```
+
+Closes the transport connection and releases any associated resources asynchronously.
+
+**Returns:**
+- A `Mono<Void>` that completes when the connection has been closed
+
+#### Close
+```java
+default void close()
```
Comment (kpavlov) [inline docs/ref-transport.md:60]: Also, would it be useful to add another variant:`close(Duration timeout)`

**Review by kpavlov — COMMENTED** [review]
Thank you, Christian, looks very nice

File: mcp/src/main/java/org/springframework/ai/mcp/server/McpServer.java (hunk 1 above)
Comment (tzolov) [reply mcp/src/main/java/org/springframework/ai/mcp/server/McpServer.java:56]: This makes sense. Should be applied to the McpClient as well.

File: docs/ref-client.md (hunk 2 above)
Comment (tzolov) [reply docs/ref-client.md:7]: You mean McpClientFactory / MpcServerFactory?

File: spring-ai-mcp-sample/src/main/resources/application.properties (hunk 3 above)
Comment (tzolov) [reply spring-ai-mcp-sample/src/main/resources/application.properties:5]: For the demo application when run into STDIO mode we should not print anything on the console in/out but the JSON RPC message part for the MCP protocol. 
Yet we can use file for the logging. 
Not sure yet what is the best way to disable the console logging in spring boot :)

**Pull request merged**